composer/callbacks/checkpoint_saver.py

```diff
@@ -0,0 +1,439 @@
+
+    Args:
+        interval (Union[str, int, Time]): The interval describing how often checkpoints should be
+            saved. If an integer, it will be assumed to be :attr:`TimeUnit.EPOCH`.
+            Otherwise, the unit must be in :attr:`TimeUnit.EPOCH` or :attr:`TimeUnit.BATCH`.
+
+            Checkpoints will be saved every ``n`` batches or epochs (depending on the unit),
+            and at the end of training.
+
+    Returns:
+        Callable[[State, Event], bool]: A function that can be passed as the ``should_save``
+            argument into :class:`CheckpointSaver`.
+    """
+    if isinstance(interval, str):
+        interval = Time.from_timestring(interval)
+    if isinstance(interval, int):
+        interval = Time(interval, TimeUnit.EPOCH)
+
+    if interval.unit == TimeUnit.EPOCH:
+        save_event = Event.EPOCH_CHECKPOINT
+    elif interval.unit == TimeUnit.BATCH:
+        save_event = Event.BATCH_CHECKPOINT
+    else:
+        raise NotImplementedError(
+            f"Unknown checkpointing interval: {interval.unit}. Must be TimeUnit.EPOCH or TimeUnit.BATCH.")
+
+    last_checkpoint_batch = None
+
+    def should_save(state: State, event: Event):
+        nonlocal last_checkpoint_batch
+        if state.get_elapsed_duration() >= 1.0:
+            # if doing batch-wise checkpointing, and we saved a checkpoint at the batch_checkpoint event
+            # right before the epoch_checkpoint event, do not save another checkpoint at the epoch_checkpoint
+            # event if the batch count didn't increase.
+            if state.timer.batch != last_checkpoint_batch:
+                last_checkpoint_batch = state.timer.batch
+                return True
+
+        if event == save_event:
+            if save_event == Event.EPOCH_CHECKPOINT and int(state.timer.epoch) % int(interval) == 0:
+                last_checkpoint_batch = state.timer.batch
+                return True
+            if save_event == Event.BATCH_CHECKPOINT and int(state.timer.batch) % int(interval) == 0:
+                last_checkpoint_batch = state.timer.batch
+                return True
+
+        return False
+
+    return should_save
+
+
+class CheckpointSaver(Callback):
+    """Callback to save checkpoints.
+
+    Example
+        >>> trainer = Trainer(..., callbacks=[
+        ...     CheckpointSaver(
+        ...         save_folder='checkpoints',
+        ...         name_format_string="ep{epoch}-ba{batch}/rank_{rank}",
+        ...         latest_symlink_format_string="latest/rank_{rank}",
+        ...         should_save="1ep",
+        ...         weights_only=False,
+        ...     )
+        ... ])
+
+    Args:
+        save_folder (str): Folder where checkpoints are saved.
+
+            If an absolute path is specified, then
+            that path will be used. Otherwise, the ``save_folder`` will be relative
+            to the folder returned by :meth:`~composer.utils.run_directory.get_run_directory`.
```

Comment:
If relative, should just be relative to cwd

Reply:
This will be addressed in #712. We will still need the run directory until we have artifact logging.